src/microsim/devices/MSDevice_GLOSA.cpp

```diff
@@ -118,37 +124,149 @@ MSDevice_GLOSA::notifyMove(SUMOTrafficObject& /*tObject*/, double oldPos,
     myDistance -= (newPos - oldPos);
     if (myNextTLSLink != nullptr && myDistance <= myRange) {
         const double vMax = myVeh.getLane()->getVehicleMaxSpeed(&myVeh);
-        const double timeToJunction = earliest_arrival(myDistance, vMax);
-        const double timeToSwitch = getTimeToSwitch(myNextTLSLink);
-#ifdef DEBUG_GLOSA
+        double timeToJunction = earliest_arrival(myDistance, vMax);
+        int countOld = 0;
+		// calculate "first" next phase, all coming Phases are calculated via "getTimeToNextSwitch"
+        double timeToSwitch = getTimeToSwitch(myNextTLSLink, countOld);
+        bool currentPhaseGreen = false;
+        bool currentPhaseStop = false;
+        bool solved = false;
+        double nextSwitch = 0;
+        nextSwitch = timeToSwitch;
+        if (myNextTLSLink->haveGreen()) { currentPhaseGreen = true; }
+        else if (myNextTLSLink->haveRed() || myNextTLSLink->haveYellow()) { currentPhaseStop = true; }
+        // else if any other phase, GLOSA does not interfere
+#if defined DEBUG_GLOSA || defined DEBUG_QUEUE
         if (DEBUG_COND) {
             std::cout << SIMTIME << " veh=" << myVeh.getID() << " d=" << myDistance << " ttJ=" << timeToJunction << " ttS=" << timeToSwitch << "\n";
         }
 #endif
-        if (myNextTLSLink->haveGreen()) {
-            myVeh.setChosenSpeedFactor(myOriginalSpeedFactor);
-            mySpeedAdviceActive = false;
-            if (timeToJunction > timeToSwitch) {
-                if (myMaxSpeedFactor > myVeh.getChosenSpeedFactor()) {
-                    const double vMax2 = vMax / myVeh.getChosenSpeedFactor() * myMaxSpeedFactor;
-                    const double timetoJunction2 = earliest_arrival(myDistance, vMax2);
-                    // reaching the signal at yellow might be sufficient
-                    const double yellowSlack = myVeh.getVehicleType().getParameter().getJMParam(SUMO_ATTR_JM_DRIVE_AFTER_YELLOW_TIME, 0);
+        if (myUseQueue) {
+		    // Detect queue length at tls
+		    double QueueLength = myNextTLSLink->getTLLogic()->getTLQueueLength(myNextTLSLink->getLaneBefore()->getID());
+            double greentime = 0;
+            double additionaljunctiontime = 0;
```

Comment:
camel case variable naming style is preferred, see https://sumo.dlr.de/docs/Developer/CodeStyle.html#variable_names

Reply:
Should I change it or will you?

Reply:
You can do it